### DISCUSSION TITLE
function setUp() behavior

### DISCUSSION THREAD

**Discussion by Tezei** [Cyfrin/foundry-full-course-cu#2737] [General]
Hiya guys,
I am wondering if when we writing tests **setUp()** function is invoked every time before each test the statement:

```
    function setUp() external {
        fundMe = new FundMe();
    }
```
is this a multiple contract deployment or this is a new instance for the contract? And is this invokes a constructor if any? I mean isn't is suitable to initialize / deploy once contract state variable once in the constructor of the test?

10x in advance!

**Comment by EngrPips** [comment]
For every test you run, a new instance of the test contract will be initiated, and the `setUp` function will act as a constructor which will construct the test contract and prepare it for the actual test function you target. If anything is still confusing, please ask questions.

**Reply by Tezei** [reply]
10x a lot! Oookay, but is there a contract deployment on each instance within setUp() function? Actually do we have a different address each time on setUp() invocation?

10x in advance!

**Reply by EngrPips** [reply]
I think we have the same address because the same parameter is always used in the creation of the test contract but maybe I am wrong.

**Reply by Tezei** [reply]
I just make a small test contract and test script attached. If I am not wrong of the test design the address remains the same of the instance.

Here is the contract:

```
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.17;

contract TedyExplore {
    function getAddress() public view returns (address) {
        // getting the contract address
        return address(this);
    }
}
```

Here is the test;
```
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

import {Test, console} from "forge-std/Test.sol";
import {TedyExplore} from "../src/TedyExplore.sol";

contract TedyExploreTest is Test {
    TedyExplore private tedyExplore;

    function setUp() public {
        tedyExplore = new TedyExplore();
    }

    function testAddress() public {
        console.log("Address: ", address(tedyExplore.getAddress()));
    }

    function testAddress_2() public {
        console.log("Address: ", address(tedyExplore.getAddress()));
    }
}
```

And the test log as follows:

Ran 2 tests for test/TedyExplore.t.sol:TedyExploreTest
[PASS] testAddress() (gas: 8697)
Logs:
  Address:  0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f

[PASS] testAddress_2() (gas: 8652)
Logs:
  Address:  0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f

**Reply by EngrPips** [reply]
Indeed, Glad you tested it.

**Reply by Tezei** [reply]
Thank you so much!

**Reply by EngrPips** [reply]
Anytime, Friend.
